---
title: "DEGs of WT and GABA mutants after submergence"
author: "Ying"
output: html_document
<<<<<<< HEAD
editor_options: 
  chunk_output_type: console
=======
>>>>>>> e29fb6740506a902bd58227d198e9adf9c6e8d6a
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<<<<<<< HEAD
```{r load libraries}
library(tidyverse)
library(magrittr)
library(edgeR)
library(ComplexHeatmap) # for UpSet Plot
```

```{r DGEList}
geneCounts <- read_delim("..//Ying_Submergence_batch1_final.txt", 
                         delim = "\t", 
                         comment = "#") %>% 
  column_to_rownames("Geneid")
colnames(geneCounts) %<>% str_remove("../3_BAM/") %>% str_remove(".STARAligned_TAIR10.sortedByCoord.bam")
# crate meta data
meta <- tibble(sample = colnames(geneCounts),
               Treatment = str_extract(sample,"cont|sub"), 
               genotype = str_extract(sample,".+(?=-(cont|sub))"),
               reps = str_extract(sample,"(?<=(cont|sub)-).+"),
               group = paste(genotype, Treatment, sep = "_") 
               # this info will fill to 'group' as col name match 'group' parameter in DGEList() 
               # and converting to a single factor Exp
)

genoLevels <- c("Col", "gad1CS", "gad1SALK", "gad21", "gad1245", "gad2OEB33", "pop28")
meta$genotype <- factor(meta$genotype, levels = genoLevels)

groupLevels <- lapply(genoLevels, function(x){paste0(x, c("_cont", "_sub"))}) %>% unlist()
meta$group <- factor(meta$group, levels = groupLevels)

# colour coding for genotype and group
genoColour <- c("#F8766D", "#7CAE00", "#00BE67", "#00BFC4", "#CD9600", "#C77CFF", "#FF61CC")
names(genoColour) <- genoLevels
groupColour <- rep(genoColour, each = 2)
names(groupColour) <- levels(meta$group)

countList <- DGEList(counts = geneCounts,
                     samples = meta) 
keep <- filterByExpr(countList)
table(keep)
countList <- countList[keep,, keep.lib.sizes=FALSE] %>% 
  calcNormFactors()

# set WT control group as reference 
countList$samples$group <- relevel(countList$samples$group, ref = "Col_cont")
```

```{r MDS plot of distances between gene expression profiles}
# Multidimensional scaling plot
# visualize sample clustering, alternatively PCA can do the job
PCoA <-
  plotMDS(countList, col=as.numeric(countList$samples$group),
          method = "logFC"
          # gene.selection = "common"  
          # "common" for PCA plot, the default is "pairwise" for PCoA plot
          )
# "pairwise" to choose the top genes separately for each pairwise comparison between the samples or "common" to select the same genes for all comparisons.

# make it easy to change appearance 
PCoA_plot <- PCoA %>%
  extract(c(5,9,10)) %>%   # coordinates info
  as.data.frame() %>%
  rownames_to_column("sample") %>%
  left_join(meta) %>%
  ggplot(aes(x = x, y = y)) +
  geom_point(aes(colour  = Treatment, shape = genotype), # shape need to be specified
             size = 3.5, alpha = 1) +
  scale_shape_manual(values = c(20,17,0,4,7,8,18)) + # set shapes
  guides(fill = guide_legend(ncol = 2)) +
  theme_bw() +
  theme(axis.text.x = element_text(vjust = -2),
        legend.position = c(0.6, 0.3),
        legend.margin = margin(-14, 0,0,0),
        text = element_text(size = 15)) + 
  labs(x = "Leading logFC dim 1 (61%)",
       y = "Leading logFC dim 2 (12%)")

```

```{r estimating dispersions}
# if with multiple factors, dispersion needs to be estimated with specified design matrix (see section 2.10.2 in edgeR user's guide updated in 21 Oct. 2019)
counts_Disp <- estimateDisp(countList)
plotBCV(counts_Disp)
```

```{r model fitting and contrast making}
# construct a design matrix with no ref group
designMat <- model.matrix(~ 0 + group, data = countList$samples)
colnames(designMat) <- levels(countList$samples$group)
# fit to model
fit <- glmQLFit(counts_Disp, design = designMat)
# make contrasts ########
myContrast <- makeContrasts(   # genes responding to submg (Ã—7 genotypes)
  wt_subVScont = Col_sub - Col_cont,     
  gad1CS_subVScont = gad1CS_sub - gad1CS_cont, 
  gad1SALK_subVScont = gad1SALK_sub - gad1SALK_cont,
  gad21_subVScont = gad21_sub - gad21_cont, 
  gad1245_subVScont = gad1245_sub - gad1245_cont,
  gad2OEB33_subVScont = gad2OEB33_sub - gad2OEB33_cont, 
  pop28_subVScont = pop28_sub - pop28_cont,
  levels = designMat)
```

```{r DE genes}
# --- a list of 7 pairs of comparisons --- #
# here DEGs not filtered in order to do volcano plot
DEGs <- lapply(colnames(myContrast), function(c){
  qlf <- glmQLFTest(fit, contrast = myContrast[,c])
  qlf$table %>% rownames_to_column("GeneID") %>% 
    mutate(adj.p = p.adjust(PValue, method = "BH"),
           comparison = c)
}) 
names(DEGs) <- colnames(myContrast)

# number of up/down regulated genes, before filtering
countDEGs <- lapply(colnames(myContrast), function(i){
  qlf <- glmQLFTest(fit, contrast = myContrast[,i])
  df <- as.data.frame(summary(decideTests(qlf))) %>% column_to_rownames("Var1") %>% t()
  df <- as.data.frame(df) %>% mutate(Genotype = i) 
  df[2,]
}) %>% 
  bind_rows() %>% 
  dplyr::select(c(Genotype, Up, Down, NotSig)) %>% 
  mutate(Total = as.numeric(Up) + as.numeric(Down) + as.numeric(NotSig))
write.csv(countDEGs, "countgenes_SubVsCont.csv")

# volcano plot: 3 by 2 facetwrap
lapply(names(DEGs), function(g){  # g: genotype_WT, the comparison
  DEGs[[g]] %>% dplyr::select(logFC, adj.p, comparison) %>% 
    mutate(Expr = ifelse(adj.p < 0.05 & abs(logFC) >=1, 
                         ifelse(logFC > 1, "Up", "Down"),
                         "NotSig"))}) %>% bind_rows() %>% as.data.frame() %>% 
  rstatix::reorder_levels("comparison", order = names(DEGs)) %>% 
    ggplot(aes(x = logFC,
               y = -log10(adj.p),
               colour = ordered(Expr, 
                                levels = c("Up", "Down", "NotSig"))))+
    geom_point(alpha = 0.2, size = 3, shape=20) +
    scale_color_manual(values = c("red", "blue", "grey")) +
    geom_vline(xintercept = c(-1, 1),
               lty = 4, colour = "#4a452a", lwd = 0.4) +
    geom_hline(yintercept = -log10(0.05),
               lty = 4, colour = "#4a452a", lwd = 0.4) +
    labs(x = "log2 (fold change)",
         y = "- log10 (adjust.p-value)") +
    theme_bw() +
    theme(legend.position = c(0.75,0.1),
          legend.title = element_blank()) +
  facet_wrap(vars(comparison), nrow = 4, scales = "fixed")

```

```{r DE genes filtered for UpSet/GO/KEGG}
DEGs2go <- lapply(names(DEGs), function(g){
  DEGs[[g]] %>% 
    filter(adj.p < 0.05 & abs(logFC) > 1) %>% 
    mutate(Expr = ifelse(logFC > 1, "Up", "Down"))
}) 
names(DEGs2go) <- names(DEGs)

# number of up/down regulated genes after filtering
countDEGs2go <- lapply(names(DEGs2go), function(g){
  Up = as.numeric(sum(DEGs2go[[g]]$Expr == "Up"))
  Down = as.numeric(sum(DEGs2go[[g]]$Expr == "Down"))
  Total = Up+Down
  Mutant = g
  df <- data.frame(Mutant, Up, Down, Total)
}) %>% 
  bind_rows()

```

```{UpSet plot for filtered DEGs: DEGs2go}
# get gene list for filtered DEGs (from DEGs2go)
geneList <- lapply(names(DEGs2go), function(g){
  DEGs2go[[g]]$GeneID
  }) 
length(geneList %>% unlist() %>% unique()) # in total 10874 
names(geneList) <- c("WT","gad1*", "gad1KO", "gad2-1","gad1245", "gad2OE","pop2-8")

# upset plot --- #
m <- make_comb_mat(geneList)
m <- m[comb_size(m) > 50] # only show groups with 50 or more genes

ss <- set_size(m)
cs <- comb_size(m)
od = order(comb_degree(m),-cs)
UpSet(m, 
      pt_size = unit(3, "mm"),
      lwd = 1, 
      set_order = order(ss),
      comb_order = od,
      comb_col = "Purple",
      
      top_annotation = HeatmapAnnotation(
        "DEGs Intersections" =  
          anno_barplot(cs,
                       ylim = c(0, max(cs)*1.1),
                       border = FALSE, 
                       gp = gpar(fill = "#4a452a"), 
                       height = unit(5, "cm")
                       ),
        annotation_name_side = "left",
        annotation_name_rot = 90),
      
      right_annotation = rowAnnotation(
        "No. of DEGs" = 
          anno_barplot(ss, 
                       baseline = 500,
                       axis_param = list(
                         # at = c(4000, 6000, 8000), # for DEGs1_go
                         at = c(500, 2000, 4000), # for DEGs2_go
                         side = "top",
                         # labels = c(4000, 6000, 8000), # for DEGs1_go
                         labels = c(500, 2000, 4000), # for DEGs2_go
                         labels_rot = 45),
                       border = FALSE, #bar_width = 0.5, extend = 0.01,
                       gp = gpar(fill = "#4a452a"), 
                       width = unit(3, "cm")),
        annotation_name_side = "top",
        annotation_name_rot = 0),
      
      row_names_side = "left",
      show_row_names = TRUE,
      row_names_gp = grid::gpar(fontsize = 10)) 
# attach number of DEGs in each group
decorate_annotation("DEGs Intersections", {
    grid.text(cs[od], x = seq_along(cs), y = unit(cs[od], "native") + unit(3, "pt"), 
        default.units = "native", just = c("left", "bottom"), 
        gp = gpar(fontsize = 7.5, col = "#4a452a"), rot = 60)
})
```

```{r save DEGs and CountList to _R directory}
DEGs2go %>% saveRDS(file.path("_R", "DEGsSub.rds"))
countList %>% saveRDS(file.path("_R", "countList.rds"))

```
=======
>>>>>>> e29fb6740506a902bd58227d198e9adf9c6e8d6a
